src/search_stream.rs

```diff
@@ -203,6 +213,9 @@ impl<'a, R: io::Read, W: Terminal + Send> Searcher<'a, R, W> {
                 break;
             }
             while self.inp.pos < self.inp.lastnl {
+                if self.opts.files_with_matches && self.match_count > 0 {
+                    break;
```

Comment:
`match_count` could increase by more than than you might expect if you're doing `inverted` matching. For regular searching, `match_count` should only be incremented by `1` at most on each iteration.

I would probably move this check to right after the call to `read_match`. i.e., `if self.opts.files_with_matches && matched { break; }`.